request-manager/spec/node-check.go

```diff
@@ -0,0 +1,498 @@
+		}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsOnceNodeCheck struct{}
+
+/* Args cannot be expected multiple times by one job. */
+func (check ArgsOnceNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	seen := map[string]bool{}
+	values := map[string]bool{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Expected == nil {
+			continue
+		}
+		if seen[*nodeSet.Expected] {
+			values[*nodeSet.Expected] = true
+		}
+		seen[*nodeSet.Expected] = true
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequenceName, &node.Name, "args -> expected", stringSetToArray(values), ""}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsNotRenamedNodeCheck struct{}
+
+/* A single arg cannot be renamed and used as multiple inputs to a node. */
+func (check ArgsNotRenamedNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	aliases := map[string]string{}
+	values := map[string]bool{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Given == nil || nodeSet.Expected == nil {
+			continue
+		}
+		if alias, ok := aliases[*nodeSet.Given]; ok && alias != *nodeSet.Expected {
+			values[*nodeSet.Given] = true
+		}
+		aliases[*nodeSet.Given] = *nodeSet.Expected
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequenceName, &node.Name, "args -> given", stringSetToArray(values), "note that if `given` is not explicitly specified, then its value is the same as `expected`"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsEachOnceNodeCheck struct{}
+
+/* Each cannot alis to job arg listed in `args`. */
+func (check ArgsEachOnceNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	seen := map[string]bool{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Expected == nil {
+			continue
+		}
+		seen[*nodeSet.Expected] = true
+	}
+
+	values := map[string]bool{}
+	for _, each := range node.Each {
+		p := strings.Split(each, ":")
+		if len(p) != 2 {
+			continue
+		}
+		if seen[p[1]] {
+			values[p[1]] = true
+		}
+		seen[p[1]] = true
+	}
+
+	if len(values) > 0 {
+		return DuplicateValueError{sequenceName, &node.Name, "each", stringSetToArray(values), "in `each: arg:alias`, a given `alias` should not be repeated in `args -> expected`"}
+	}
+
+	return nil
+}
+
+/* ========================================================================== */
+type ArgsEachNotRenamedNodeCheck struct{}
+
+/* Each cannot rename job args listed in `args`. */
+func (check ArgsEachNotRenamedNodeCheck) CheckNode(sequenceName string, node NodeSpec) error {
+	aliases := map[string]string{}
+	for _, nodeSet := range node.Args {
+		if nodeSet.Given == nil || nodeSet.Expected == nil {
+			continue
+		}
+		// This is fine:
+		// each: [arg:alias]
+		// args:
+		//   - expected: arg
```

Comment:
Well, one reason is that the second case appears in either the docs or our actual specs. (I don't remember right now which it is.)

But also, the second case isn't really renaming the arg--the operation it performs is something like `nodes --> nodes -->> node` (where `-->` is the identity operation, and `-->>` is my attempt to represent a "split" operation). The first case does `nodes --> nodeList` _and_ `nodes -->> node`. So they really are different operations. Whether or not we want to warn in the second case is another matter, but even if we do choose to warn, it'd probably be best as a different check.

Reply:
`the second case appears in either the docs or our actual specs` oop. I think there used to be a bug where you (only sometimes, weirdly) had to include the list arg in the `args`, so perhaps it's a relic of that?

In the first case, the result is that the job gets both `node` and `nodeList` (which started as `nodes`) passed in as args. In the second case, it gets both `node` and `nodes` as an arg. Other than the name of the "list" arg, there doesn't really seem to be a difference to me. Using your terms, the second case basically does `nodes --> nodes` (a no-op) and `nodes -->> node`.

I agree in the second case it's not exactly _renaming_ the arg (and in the first case it's more `args`, rather than `each` that's doing "renaming", hence I don't like `ArgsNotRenamedByEach`), but I think the solution is to change the name of the check.